### ISSUE TITLE
Tasks fail with error "Unable to complete atomic operation, key modified"

### ISSUE THREAD

**Issue by mavenugo** [moby/moby#35310]
This is a spin-off of #35290 to address the error `Unable to complete atomic operation, key modified` that causes the Tasks to be scheduled on the node.

```
Oct 26 09:43:13 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:13.534861035-07:00" level=error msg="fatal task error" error="Unable to complete atomic operation, key modified" module=node/agent/taskmanager node.id=mv4e72vawng0s82vvnb0tatc7 service.id=m9484e8pmri988v4n3bossl0x task.id=c1qap6z4nr50k9cacus63d8is
```

Thanks to @abhi , we reproduced the issue with debug enabled. The root-cause of this seems to be 
introduced via https://github.com/moby/moby/pull/34674 

Looking at the attached logs, 

```
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806458067-07:00" level=debug msg="releasing IPv4 pools from network orange (pz8tw8vw40ogxi01c3pcthldi)"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806483219-07:00" level=debug msg="ReleaseAddress(LocalDefault/10.0.4.0/24, 10.0.4.1)"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806504836-07:00" level=debug msg="ReleasePool(LocalDefault/10.0.4.0/24)"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806643570-07:00" level=debug msg="moby-demo-ubuntu-1(48d7c3d0f4a7): leaving network pz8tw8vw40ogxi01c3pcthldi"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806657694-07:00" level=debug msg="cleanupServiceDiscovery for network:pz8tw8vw40ogxi01c3pcthldi"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806665552-07:00" level=debug msg="cleanupServiceBindings for pz8tw8vw40ogxi01c3pcthldi"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806890107-07:00" level=warning msg="driver error deleting endpoint orange-endpoint : network id \"pz8tw8vw40ogxi01c3pcthldi\" not found"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806904850-07:00" level=debug msg="Releasing addresses for endpoint orange-endpoint's interface on network orange"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806915275-07:00" level=debug msg="ReleaseAddress(LocalDefault/10.0.4.0/24, 10.0.4.16)"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806935796-07:00" level=warning msg="Failed to release ip address 10.0.4.16 on delete of endpoint orange-endpoint (d68ba4cb8deea9ed20c05c0abbd19c4db3250824dbc9133afc5bf2977f8e4765): cannot find address pool for poolID:LocalDefault/10.0.4.0/24"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.806986817-07:00" level=error msg="network orange remove failed: unknown network orange id pz8tw8vw40ogxi01c3pcthldi" module=node/agent node.id=mv4e72vawng0s82vvnb0tatc7
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.807004206-07:00" level=error msg="remove task failed" error="unknown network orange id pz8tw8vw40ogxi01c3pcthldi" module=node/agent node.id=mv4e72vawng0s82vvnb0tatc7 task.id=d675fekf2hc5vxx188d5c0r5e
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.807146117-07:00" level=warning msg="driver error deleting endpoint orange-endpoint : network id \"pz8tw8vw40ogxi01c3pcthldi\" not found"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.807160203-07:00" level=debug msg="Releasing addresses for endpoint orange-endpoint's interface on network orange"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.807169786-07:00" level=debug msg="ReleaseAddress(LocalDefault/10.0.4.0/24, 10.0.4.16)"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.807182810-07:00" level=warning msg="Failed to release ip address 10.0.4.16 on delete of endpoint orange-endpoint (d68ba4cb8deea9ed20c05c0abbd19c4db3250824dbc9133afc5bf2977f8e4765): cannot find address pool for poolID:LocalDefault/10.0.4.0/24"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.809217709-07:00" level=error msg="network orange remove failed: unknown network orange id pz8tw8vw40ogxi01c3pcthldi" module=node/agent node.id=mv4e72vawng0s82vvnb0tatc7
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.809233253-07:00" level=error msg="remove task failed" error="unknown network orange id pz8tw8vw40ogxi01c3pcthldi" module=node/agent node.id=mv4e72vawng0s82vvnb0tatc7 task.id=qofrrwn7nf917tq9s93es6s8h
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.809383599-07:00" level=warning msg="driver error deleting endpoint orange-endpoint : network id \"pz8tw8vw40ogxi01c3pcthldi\" not found"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.809398355-07:00" level=debug msg="Releasing addresses for endpoint orange-endpoint's interface on network orange"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.809408880-07:00" level=debug msg="ReleaseAddress(LocalDefault/10.0.4.0/24, 10.0.4.16)"
Oct 26 09:43:11 moby-demo-ubuntu-1 dockerd[14670]: time="2017-10-26T09:43:11.809421881-07:00" level=warning msg="Failed to release ip address 10.0.4.16 on delete of endpoint orange-endpoint (d68ba4cb8deea9ed20c05c0abbd19c4db3250824dbc9133afc5bf2977f8e4765): cannot find address pool for poolID:LocalDefault/10.0.4.0/24"

```

It seems that even after the network is cleaned up on a worker node, the Load-balancer endpoint and sandbox for that network is continuously tried to be cleaned up and keeps failing.
Because of this, the `endpoint_cnt` key is created on the store while handling the endpoint cleanup phase. When the network is created again, it fails because of existing key.

We could hack endpoint_cnt to prevent this failure. But I think there is a fundamental issue with the way the per-network-per-node LB endpoint and sandbox are managed. We must resolve this asap.

Should we back-off #34674  till the issue is properly addressed ?
cc @pradipd @vieux @fcrisciani @abhi 

[daemon.log](https://github.com/moby/moby/files/1420252/daemon.log)

**Comment by ZigmunD** [comment]
Same here. Docker 17.10, 5 managers, 2 workers, Proxmox VMs, Ubuntu 16.04.3.
After another service update I got:
```
zigmund@docker-m1.alahd.kz.prod:~$ docker service ps stack-ms-comment_kolesa --no-trunc 
ID                          NAME                            IMAGE                                                                                                                               NODE                      DESIRED STATE       CURRENT STATE             ERROR                                                 PORTS
ufdnbnkc0n2urcjsx15otblm3   stack-ms-comment_kolesa.1       private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w2.alahd.kz.prod   Ready               Rejected 2 seconds ago    "Unable to complete atomic operation, key modified"   
bxtj1rgv04r7gjan4ydmzjg5l    \_ stack-ms-comment_kolesa.1   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w2.alahd.kz.prod   Shutdown            Rejected 7 seconds ago    "Unable to complete atomic operation, key modified"   
zzc3edtkioz53p92t9353cy8o    \_ stack-ms-comment_kolesa.1   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w2.alahd.kz.prod   Shutdown            Rejected 25 minutes ago   "Unable to complete atomic operation, key modified"   
zzq4rar1ie5v0gjmkmv12jpfx    \_ stack-ms-comment_kolesa.1   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w1.alahd.kz.prod   Shutdown            Rejected 3 hours ago      "Unable to complete atomic operation, key modified"   
zz7rq3qi5x7h8coygexan2ncy    \_ stack-ms-comment_kolesa.1   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w1.alahd.kz.prod   Shutdown            Rejected 4 hours ago      "Unable to complete atomic operation, key modified"   
xfkk0l0b5ey4ppva8t5qicbvb    \_ stack-ms-comment_kolesa.1   private-registry:5000/ms/comment:r46                                                                           docker-w2.alahd.kz.prod   Shutdown            Running 20 hours ago                                                            
w9bd33tulx3t5xwq0y0wlhzsu   stack-ms-comment_kolesa.2       private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w1.alahd.kz.prod   Ready               Rejected 1 second ago     "Unable to complete atomic operation, key modified"   
b72hb4a3vcyqne0weaobazz5u    \_ stack-ms-comment_kolesa.2   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w1.alahd.kz.prod   Shutdown            Rejected 6 seconds ago    "Unable to complete atomic operation, key modified"   
zze8pinm4rw0cuhwtre9lniyx    \_ stack-ms-comment_kolesa.2   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w1.alahd.kz.prod   Shutdown            Rejected 10 minutes ago   "Unable to complete atomic operation, key modified"   
zxai9t2lu5kfetma2ronkb6nz    \_ stack-ms-comment_kolesa.2   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w1.alahd.kz.prod   Shutdown            Rejected 19 minutes ago   "Unable to complete atomic operation, key modified"   
zxot2qip81vtboxudkns1k3d8    \_ stack-ms-comment_kolesa.2   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w1.alahd.kz.prod   Shutdown            Rejected 2 hours ago      "Unable to complete atomic operation, key modified"   
zytwajrywy0unenpogur4lp1w    \_ stack-ms-comment_kolesa.2   private-registry:5000/ms/comment:r51@sha256:74fcc78a4a615d5be02e7e02b8618c865bb58eee73d85e00301c6520584f1b7b   docker-w1.alahd.kz.prod   Shutdown            Rejected 3 hours ago      "Unable to complete atomic operation, key modified"
```

Swarm completely unusable at the moment, we cannot update services.
Have not seen this error before 17.10.

**Comment by ZigmunD** [comment]
As workaround that works for me - create independent overlay network and use it as external in stacks.

**Comment by thaJeztah** [comment]
This one also looks to be related to the ILB/ELB PR; https://github.com/moby/moby/issues/35281

**Comment by pradipd** [comment]
FYI @msabansal @dineshgovindasamy

**Comment by pradipd** [comment]
@theJeztah What commit is docker 17.10-ce build from?  Is there a tag?

**Comment by pradipd** [comment]
@abhi Can you share the repro steps?

**Comment by thaJeztah** [comment]
@pradipd see https://github.com/docker/docker-ce/releases/tag/v17.10.0-ce

**Comment by pradipd** [comment]
I sync'd up with @abhi and @mavenugo and got all the relevant details.
I'm investigating now. Thanks!

**Comment by luckygoodness** [comment]
yea why not how do you want it

**Comment by vieux** [comment]
@pradipd any news on this ?

**Comment by omerh** [comment]
will this fix be on 17.11 ?

**Comment by vieux** [comment]
@omerh yes, either this will be fixed or the commit that introduced the issue will be reverted

**Comment by pradipd** [comment]
@vieux  
Making progress.  
I can't repro at all on windows.
I can repro this 100% on linux, which seems odd if this is a timing issue.
the issue repros for me with the following:
docker swarm init --advertise-addr=<ip>
docker network create -d overlay --subnet 192.168.1.0/24 --gateway 192.168.1.1 o1
docker service create --name s1 --replicas 3 --network o1 --publish 8080:80 nginx
docker service rm s1
docker service create --name s1 --replicas 3 --network o1 --publish 8080:80 nginx

I'm a little unsure as how I didn't see this before and/or tests didn't catch this.  
On the bright side, with a fairly consistent repro, I should be able to fix.

**Comment by svscorp** [comment]
> As workaround that works for me - create independent overlay network and use it as external in stacks.

@ZigmunD for me it is still the case even with independent overlay or/and ingress networks, unfortunately.

**Comment by ZigmunD** [comment]
@svscorp Haven't changed any default networks.

**Comment by fvilers** [comment]
I have the same error message when scaling up a service on a Raspberry PI swarm.

**Comment by kassanmoor** [comment]
any way to safely downgrade docker? this issue is annoying :|

**Comment by kassanmoor** [comment]
Ant new in app this? apparently a workaround is add a new manager with a lower docker version sadly I was unable to test further since was a windows manager that get unresponsive after run a service :|

**Comment by pradipd** [comment]
Sync'd up with @mavenugo  and we have a fix.  However, we will bring the fix into 1711.
I think the plan is to revert this change in 17.10.  @mavenugo @vieux please confirm that is the plan.

**Comment by vieux** [comment]
@pradipd we will release 17.11 very soon and 17.10 was only a edge release, so I think fixing it in 17.11 is enough ? do you have a timeline for the fix ? (RC3 is tentatively scheduled for monday)

**Comment by pradipd** [comment]
@vieux Should send a PR out by EOD.

**Comment by swaroopkundeti** [comment]
@pradipd Any news on this?

**Comment by fentas** [comment]
@swaroopkundeti see release  notes https://github.com/docker/docker-ce/releases
> Fix tasks fail with error "Unable to complete atomic operation, key modified" docker/libnetwork#2004

**Comment by pradipd** [comment]
PR's are out in both libnetwork and moby repo's.  Will ping @mavenugo and @vieux to check the status.

**Comment by kassanmoor** [comment]
so, this will be fixed on  17.11? when will be released?

**Comment by parrisvarney** [comment]
If anybody is stuck, I've had success working around this issue by deleting the stack and network, then recreating both.

```sh
docker stack rm my_stack
docker network rm my_network
docker network create --driver overlay my_network
docker stack deploy my_stack
```

**Comment by kassanmoor** [comment]
@parrisvarney great, but is not a good option if you have several services in the network :( waiting for final fix

**Comment by DBLaci** [comment]
@parrisvarney thank you for the workaround but I must mention this issue occurs on totally new swarm cluster with docker stack deploy from scratch and redeploy or additional services can cause it again. Removing stack and readding is good for a temporary workaround in development environment only.

**Comment by parrisvarney** [comment]
@kassanmoor @DBLaci  Agreed, it's not sustainable or even practical.  I have personally been able to move forward with Docker dependent development for the time being by nuking the network, so figured I'd share.

**Comment by kassanmoor** [comment]
any update here? I am considering downgrade as an option since I did not see if will be fixed on 17.11 and also do not know when will new version will be available

**Comment by stevelacy** [comment]
I downgraded to 17.09 as not a single service would start on 17.10.

**Comment by shankarkc** [comment]
Its really P0 defect for people using service. Can you guys provide a fix urgently?

**Comment by pradipd** [comment]
Ping @vieux @mavenugo

**Comment by aruzamen** [comment]
Is there any news on this issue, waiting for a fix.

**Comment by mavenugo** [comment]
A workaround for this issue is pushed in libnetwork and the vendoring is stuck in https://github.com/moby/moby/pull/35415. That will workaround the issue.
But the actual issue is being addressed by @pradipd in his PRs that are under code-review.

**Comment by LRNZ09** [comment]
I'm also having this issue using Traefik with `Docker version 17.10.0-ce, build f4ffd2511c`

**Comment by kassanmoor** [comment]
Docker 17.11 is out, the fix is included there?

**Comment by omerh** [comment]
According to the release node it did
https://github.com/docker/docker-ce/releases/tag/v17.11.0-ce

Fix tasks fail with error "Unable to complete atomic operation, key modified" [docker/libnetwork#2004](https://github.com/docker/libnetwork/pull/2004)

Didn't checked it yet on my clusters

**Comment by morph027** [comment]
Just spinning up a test cluster...will report when done

**Comment by morph027** [comment]
Looks good so far

* created an overlay network `proxy`
* deployed docker flow proxy stack
* deployed swarm visualizer stack
* scale proxy to 3

![bildschirmfoto vom 2017-11-23 10-33-29](https://user-images.githubusercontent.com/600106/33166247-e3317cee-d039-11e7-8979-d4349611a79a.png)

**Comment by DBLaci** [comment]
I confirm it's working with "Docker version 17.11.0-ce, build 1caf76c" - thank you.

**Comment by morph027** [comment]
Just upgraded my staging swarm (runs services for CI) and it works! Thank you guys!

**Comment by shankarkc** [comment]
Cool thanks guys for updating that fix is working

**Comment by svscorp** [comment]
Do I need to recreate networks or what is only required is install new version on all the nodes?restart daemon/VMs?

**Comment by svscorp** [comment]
Rolled back to 17.07 and my stack is stable again.

**Comment by shankarkc** [comment]
Yes. I too noticed all containers in swarm were dead after some time. I
rolled back to 17.05 and then my service came back online again.

On Nov 24, 2017 3:36 PM, "Ilia Shakitko" <notifications@github.com> wrote:

I have updated the cluster to 17.11. Issues:

   1. Just updating docker engine from 17.07 to 17.11 -> broke the swarm
   and services (thankfully data is persisted on the disk)
   2. After recreating Swarm, 50% VMs have lost their 'eth1' it was renamed
   to 'veth...'
   3. (not sure it is related, but it *wasn't the case* before) After swarm
   and docker stack was recreated it worked for a while stable, but then all
   the tools are occasionally hanging in their connection to LDAP (which is
   not part of the docker services). After ~1h of running or more, something
   like that.
   4. Enormous amount of *(10-20 per second!!!)* dockerd[25583]:
   time="2017-11-24T11:03:14.756972678+01:00" level=info msg="Node join
   event ..." - it beings enormous IO that results in costs for
   storage/performance + issues because disk is overloaded 100% of the time on
   every node. What can be done immediately about this?

No "atomic operation failed" issue spotted so far, but things are unstable
:(

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/moby/moby/issues/35310#issuecomment-346790709>, or mute
the thread

**Comment by svscorp** [comment]
I like people reacting with "confused" emoji. The "most useful" feedback/reaction on the issue :)

To the "confused" people - please elaborate on your confusion? Otherwise maintainers or someone who COULD actually help will just skip mine and @shankarkc report leaving without attention (like... yeah, so many ":\" evaluations received).

**Comment by eXistenZNL** [comment]
@svscorp are you suggesting we should not use this emoticon anymore? Or should we skip your comment without attention? I'm confused...

**Comment by stevelacy** [comment]
Now I'm starting to go off topic...

@svscorp I use the :confused: emoticon as it is the closest to an unhappy icon, for instance responding in kind to your: "In an hour or so after restarting services (or docker engine) problems arise. ..... have to reboot again".
GH does not have a `:(` response icon, the closest visually is the `:confused:` icon

**Comment by svscorp** [comment]
@eXistenZNL @stevelacy okay, lets indeed kill the off-top right here as it floods the conversation.

No offense please, I was trying to understand is there maybe any useful info / question behind the emoji. Is it "just confused / nothing to add or ask" or "confused as don't like / agree with the comment". ExistenZNL, in given 2 choices smiley is at least some attention :)
----
UPD (meanwhile): had conversation with the team and my next steps would be to isolate communication between services and service-to-external-tools connection issues. I'll share any outcome once have it.

**Comment by eXistenZNL** [comment]
@svscorp no offence taken buddy :taco: :confused: :confetti_ball: 

Please ignore this comment so we can focus productively on the original issue in this ticket :point_up_2:

**Comment by thaJeztah** [comment]
This is resolved through https://github.com/moby/moby/pull/35422, which is merged in master and will be included in the next Docker release as well

**Comment by shurshun** [comment]
+ the same problem

**Comment by fcrisciani** [comment]
@shurshun which version?

**Comment by shurshun** [comment]
@fcrisciani 17.11.0-ce on CentOS 7 with 4.14.5-1 kernel

**Comment by thaJeztah** [comment]
@shurshun that version does not yet have the fix; the fix will be in the next (17.12) release

**Comment by swaroopkundeti** [comment]
@thaJeztah Any time frame for 17.12 release?

**Comment by thaJeztah** [comment]
Were still looking at some fixes we'd like to get in before releasing; hopefully _soon_

**Comment by swaroopkundeti** [comment]
@thaJeztah Hoping this is fixed in 17.09.1-ce (2017-12-07)

**Comment by thaJeztah** [comment]
@swaroopkundeti the problem being discussed here was never in 17.09, so nothing to fix in that release; see the first comment;

> The root-cause of this seems to be introduced via #34674

That pull request went into Docker 17.10, so Docker 17.09 was not affected.

But 17.12 is now released, so either 17.09 or 17.12 should work
